### PR TITLE
feat(trainer): pack RL multimodal samples

### PR DESCRIPTION
**Pull request by hubert-marek**
## Summary
- Adds RL `pack_multimodal` support for eager multimodal samples while preserving run/LoRA and rank modality scheduling constraints.
- Makes `seq_lens` the first-class packed sample boundary contract and threads it through transport, RL data loading, trainer forward, and Qwen3.5 MRoPE.
- Gates packed multimodal training with an explicit model capability validator, varlen attention validation, and config-layer rejection for VLM context parallelism.

## Verification
- `uv run --no-project python -m py_compile src/prime_rl/trainer/batch.py src/prime_rl/trainer/rl/data.py src/prime_rl/transport/types.py tests/unit/orchestrator/test_batch.py`
- `uv run --no-project --with ruff==0.13.0 ruff check --fix --config=pyproject.toml <PR files>`
- `uv run --no-project --with ruff==0.13.0 ruff format --config=pyproject.toml <PR files>`
- `PYTHONPATH="src:packages/prime-rl-configs/src" uv run --no-project --with pytest --with numpy --with msgspec --with torch --with jaxtyping --with tomli-w --with pydantic --with psutil --with setproctitle --with pandas --with rich --with transformers --with wandb --with 'prime-pydantic-config[toml]' --with pyzmq --with loguru python -m pytest tests/unit/orchestrator/test_batch.py -q`
- `PYTHONPATH=src:packages/prime-rl-configs/src uv run --no-project --with ruff ruff check src/prime_rl/utils/vlm.py src/prime_rl/trainer/rl/train.py packages/prime-rl-configs/src/prime_rl/configs/trainer.py tests/unit/utils/test_vlm.py tests/unit/test_configs.py`
- `PYTHONPATH=src:packages/prime-rl-configs/src uv run --no-project --with pytest --with torch --with numpy --with transformers --with psutil --with setproctitle --with loguru pytest tests/unit/utils/test_vlm.py -q`
- Direct `TrainerConfig` validation check for VLM + `cp=2` + `pack_multimodal=true`

---

> [!NOTE]
> **Medium Risk**
> Changes batch packing, multimodal tensor layout, and MRoPE/varlen attention boundaries—incorrect packing could skew RL loss or attention masks, though CP/VLM and unsupported models are blocked at config/runtime.
> 
> **Overview**
> Adds optional **packed multimodal** RL training: when `trainer.pack_multimodal` is true (default) and the model supports it, eager image/text rollouts from the same run/LoRA can share one micro-batch instead of one sample per batch.
> 
> **`seq_lens`** is introduced as the packed-segment boundary (transport → packer → dataloader → `forward`). For Qwen-style MRoPE (`image_grid_thw`), trainer `position_ids` are no longer passed; **`seq_lens`** drives varlen attention and internal 3D positions. **`mm_kwargs`** are concatenated when dtypes/shapes match.
> 
> **Gating:** `resolve_pack_multimodal` checks model `supports_packed_multimodal_training`, varlen flash attention, and rejects multimodal packing with CP. Config validation rejects **VLM + context parallelism**. Debug/CI multimodal configs set `pack_multimodal = false` explicitly.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3ee2a45a475f3ff21ab1d79593da9a98af2dc106. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/utils/vlm.py [hunk 1]
```diff
@@ -97,6 +100,38 @@ def get_final_logit_softcapping(model_config: PretrainedConfig) -> float | None:
...
+    """Return whether the model advertises safe packed multimodal training."""
+    for candidate in _iter_wrapped_modules(model):
+        supported = getattr(candidate, "supports_packed_multimodal_training", None)
+        if supported is not None:
+            return bool(supported)
+
+    return False
+
+
+def get_packed_mm_disabled_reasons(
```
Comment (samsja) [inline src/prime_rl/utils/vlm.py]: lest call this, "validate_multi_modal_pack" and lets fail instead of disabling if not compatible.

also the cp X multi modal pack should be done at the config layer

File: src/prime_rl/trainer/batch.py [hunk 2]
```diff
@@ -95,6 +95,55 @@ def _truncate_mm(
...
+        raise ValueError(
+            "EncodedTensor byte length does not match dtype and shape: "
+            f"dtype={encoded.dtype}, shape={encoded.shape}, "
+            f"data_nbytes={len(encoded.data)}, expected_nbytes={expected_nbytes}"
+        )
+
+
+def _append_encoded_tensor(dst: EncodedTensor, src: EncodedTensor, key: str) -> None:
+    _validate_encoded_tensor_payload(dst)
+    _validate_encoded_tensor_payload(src)
```
Comment (samsja) [inline src/prime_rl/trainer/batch.py]: do we really need this check here ?

File: src/prime_rl/trainer/batch.py [hunk 3]
```diff
@@ -95,6 +95,55 @@ def _truncate_mm(
...
+    if dst.dtype != src.dtype:
+        raise ValueError(f"Cannot pack mm_kwargs[{key!r}] with different dtypes: {dst.dtype} vs {src.dtype}")
+    if len(dst.shape) == 0 or len(dst.shape) != len(src.shape) or dst.shape[1:] != src.shape[1:]:
+        raise ValueError(f"Cannot pack mm_kwargs[{key!r}] with incompatible shapes: {dst.shape} vs {src.shape}")
```
Comment (samsja) [inline src/prime_rl/trainer/batch.py]: same here can't we check before ?

File: src/prime_rl/trainer/batch.py [hunk 4]
```diff
@@ -95,6 +95,55 @@ def _truncate_mm(
...
+        return False
+    return all(
+        dst[key].dtype == src[key].dtype
+        and len(dst[key].shape) > 0
+        and len(dst[key].shape) == len(src[key].shape)
+        and dst[key].shape[1:] == src[key].shape[1:]
+        for key in dst
+    )
+
+
```
Comment (samsja) [inline src/prime_rl/trainer/batch.py]: not the biggest fan here, overly defensive code there must be an easier way to do this

File: src/prime_rl/trainer/batch.py [hunk 5]
```diff
@@ -95,6 +95,55 @@ def _truncate_mm(
...
+def _has_video_tokens(sample: MicroBatch) -> bool:
+    return sample.mm_token_type_ids is not None and 2 in sample.mm_token_type_ids
```
Comment (samsja) [inline src/prime_rl/trainer/batch.py]: are we working with video ? if not lets just remvoe this

File: src/prime_rl/trainer/batch.py (hunk 2 above)
Comment (hubert-marek) [reply src/prime_rl/trainer/batch.py]: We can remove them

File: src/prime_rl/utils/vlm.py [hunk 6]
```diff
@@ -97,6 +100,26 @@ def get_final_logit_softcapping(model_config: PretrainedConfig) -> float | None:
     return getattr(model_config.get_text_config(), "final_logit_softcapping", None)
 
 
+def supports_packed_multimodal_training(model: nn.Module) -> bool:
+    """Return whether the model advertises safe packed multimodal training."""
+    return bool(getattr(model, "supports_packed_multimodal_training", False))
```
Comment (samsja) [inline src/prime_rl/utils/vlm.py:105]: not sure if we really need this function

**Review by samsja — COMMENTED** [review]
the seq len paramters we added that is passed arround is not used for non multi modal, can we rather have mm_seq_lens that is none if not mutli modal ?
